### ISSUE TITLE
List of things needed for releasing 1.0

### ISSUE THREAD

**Issue by Etiene** [sailorproject/sailor#75]
(The design repository (https://github.com/sailorproject/design/issues) didn't really work out, so I am posting this here)

What do you think is *necessary* for claiming to be stable enough for production use so we can release a 1.0 version?

I already had some in mind so I'll list them here, please list more things as well

* I just measured tests coverage for the first time today. It is shitty (45%). I'd say it should go above 95% at the very least
(https://github.com/sailorproject/sailor/issues/68)

* We also need performance tests so we can defend ourselves technically a bit better
(https://github.com/sailorproject/sailor/issues/55)

* Improve overall compatibility with Nginx, since it is the most stable / active web server supporting Lua, it is only logical we give more support to them too
(https://github.com/sailorproject/sailor/issues/72)

* Rewrite the module for sending emails, because it is an essential feature and it is shitty if it doesn't work properly
(https://github.com/sailorproject/sailor/issues/59)

* Test the Xavante form file input upload because Xavante will be often used as a test server, so keeping up support to Xavante is ideal too
(https://github.com/sailorproject/sailor/issues/74)

* Enable users to easily develop and integrate their own extensions for Sailor, such as user-login modules etc. Example: http://www.yiiframework.com/extensions/
(https://github.com/sailorproject/sailor/issues/57)

What else should go here?

(Also, feel free to take this list as something significative to pick and contribute)

**Comment by catwell** [comment]
I was thinking Lua 5.3 compatibility. I think Sailor is not 5.3 compatible yet because of some dependencies, but do you know what is blocking it exactly?

**Comment by Etiene** [comment]
I'm not sure 5.3 compatibility is this essential because it's not compatible with openresty anyway, which is a direction I think we should take. So 5.3 compatibility would be useful for running on Xavante and I'm not sure if Xavante itself runs on 5.3, I know cgilua, doesn't and as far as I know and it's not 5.3 compliant either...

**Comment by catwell** [comment]
Xavante master branch on Redis does run on 5.3. mod_lua (Apache) is also 5.3 by default now.

It may not be important for 1.0 but increasingly what people have installed will be 5.3 and we probably want Sailor to work out of the box.

**Comment by hishamhm** [comment]
Don't be so hard on yourself for test coverage for 1.0. You have a considerable matrix of execution scenarios going. To this day, LuaRocks only gets 80% coverage (and it's not like I was really lazy about it — heck, I wrote LuaCov to coverage-test LuaRocks!). Having many codepaths that are active only in different configurations makes coverage hard.

**Comment by catwell** [comment]
Yes, regarding other topics:

- I agree with Hisham, 80% coverage is a more reasonable target.
- I don't think performance is so important at this stage. Still good to know but the performance of, say, Rails has been terrible for a long time and it did not prevent them from being successful. Also, benchmarks of Web frameworks don't really tell the whole story anyway: every application is different, it depends on deployment, etc.
- Compatibility with nginx: basically what you must do is use the `resty-*` libraries in OpenResty context like [I did with one of my own modules during FOSDEM](https://github.com/catwell/tls-mailer/commit/59f14982751965b4454ee03a9a9e73efe3c744d8) ;)
- Re. emails consider my module above ;) which I have now released on LuaRocks.
- Xavante: agreed.
- Extensions: should not block 1.0 IMO (that could be a 1.1 feature).

**Comment by Etiene** [comment]
Ok :) thanks @hishamhm 

@catwell About 5.3, ok, I need to take a look at that but I think dependencies will block :( I'm gonna try and see what happens

**Comment by Etiene** [comment]
@catwell Yes, I'm using resty library for MySQL when resty context is around. I need to do that for the rest now

**Comment by Etiene** [comment]
Oh and I'm not thinking about optimising performance or pay attention to it, just measuring it... ^^'

**Comment by Etiene** [comment]
This pretty awesome PR made me realise coverage was not that bad! (https://github.com/sailorproject/sailor/pull/76) Thanks @mpeterv 👯
We should still increase it a little bit, though ^_^

**Comment by hishamhm** [comment]
+1 for Lua 5.3 compat and Xavante support from me. I'm pinging Tomas for a new Cgilua release since current master has the fixes merged in.

**Comment by mpeterv** [comment]
@hishamhm Xavante and wsapi also need releases (fixes were merged in December).

**Comment by mpeterv** [comment]
Probably copas, too, with  keplerproject/copas#40. And maybe coxpcall too? That's a lot of releasing...

**Comment by hishamhm** [comment]
@mpeterv Yes. I pinged @tomasguisasola already wrt Cgilua. Once that is up, I'll ping @mascarenhas for Xavante. Copas is being maintained by @Tieske. WSAPI is @mascarenhas too.

**Comment by hishamhm** [comment]
I can do coxpcall myself.

**Comment by hishamhm** [comment]
coxpcall 1.16.0 released!

**Comment by Etiene** [comment]
That's efficiency right here! :D Thanks @hishamhm!! :dancer: 
Now only cgilua to go!

**Comment by Etiene** [comment]
Does everyone think extensions should be at later versions, as @catwell suggested? Or do you think it should be at 1.0?

**Comment by xspager** [comment]
I would like to suggest instead of focusing in any webserver maybe the focus should be getting everything working on Xavante for development and with fastcgi so it wouldn't matter what webserver is used as long it support fastcgi.

**Comment by xspager** [comment]
Documentation (using LDoc) would be nice to have.

**Comment by mascarenhas** [comment]
@catwell, do you mean GitHub instead of Redis? The master branch at keplerproject/xavante has a couple of commits from @mpeterv that added Lua 5.3 compatibility. He also updated keplerproject/cgilua and keplerproject/wsapi at the same time, so if everything is ok it is just a matter of doing new releases.

**Comment by catwell** [comment]
@mascarenhas Yes I did! I edited my comment to avoid further confusion. Thanks.

**Comment by Etiene** [comment]
Concerning Lua5.3 compatibility, the issue is that, apart from postgres, the LuaSQL library is not fully compatible with it. So, apparently, sailor itself runs well on xavante & 5.3 (see here: https://travis-ci.org/sailorproject/sailor/builds/108734493). The LuaSQL module is not a dependency, though, but will be most likely used, so I'm not sure how to proceed now. Should I state we're compatible and make a document with a compat table?

**Comment by catwell** [comment]
We should probably try to make LuaSQL 5.3-compatible :)

**Comment by Etiene** [comment]
That sounds a bit too complicated for me but I can try :~

Concerning performance tests: this is not necessary for having a stable sailor, of course, but I think we absolutely should have it for the release announce "Look, we have this software and look at how fast it is!!! :D" No matter what we will get, my instinct says it will be good overall just because of Lua itself. So we won't need to tweak anything, just measure it.

Also, can we take a moment to talk about LuaSec? I removed the email module long ago with the goal of adding it as a separate external and optional module only because of dependency on LuaSec. Installing LuaSec through LuaRocks almost never will work at first try, it will never find the required headers, and I thought this was unnecessary pain for people installing sailor. I wouldn't have an issue getting the email module back and add LuaSec back as a dependency if this was better...

**Comment by catwell** [comment]
Re. LuaSec you just mean he has to do a release right?

Also, the email module does not depend on LuaSec when using OpenResty. But LuaRocks currently has no way to deal with this kind of dependency :(

**Comment by Etiene** [comment]
Wait. Nevermind. Either I was using an old version of LuaRocks or an old version of LuaSec. I'm not having that issue anymore (even on LuaSec 0.5)! Sorry for bugging people ^^' I'm not sure about it's 5.3 compatibility though. Gonna check that now

**Comment by Etiene** [comment]
Ok! Apparently that's alright on 5.3 too. I guess I only need to write a switche between resty's mail and my mail module according to server used! Yay

**Comment by Etiene** [comment]
Would anyone like to contribute to adding Sailor here? https://www.techempower.com/benchmarks/

**Comment by brunoos** [comment]
LuaSec 0.5 has some issues with Lua 5.3.
Next version will solve it (release is coming soon)

**Comment by mpeterv** [comment]
On master branch luasql-mysql works with Lua 5.3 (at the very least it doesn't crash on `require`), so, again, it just needs a release.

**Comment by Tieske** [comment]
copas 2.0.1 has been released

**Comment by Etiene** [comment]
Hey, I know you said you think extensions is not really necessary for the 1.0, but I'd still like your input!
I've created a separate issue for this: https://github.com/sailorproject/sailor/issues/82

**Comment by xspager** [comment]
What about updating Bootstrap?

**Comment by Etiene** [comment]
Good idea!

**Comment by felipedaragon** [comment]
I suggest #58 to the list

**Comment by Etiene** [comment]
Added!
